### DISCUSSION TITLE
Playing sound from balance bot

### DISCUSSION THREAD

**Discussion by dsplaisted** [pybricks/support#1774] [Q&A]
Hi,

I've modified the balance bot code found [here](https://github.com/pybricks/pybricks-projects/blob/master/sets/mindstorms-robot-inventor/other-models/balancer/main.py) to let the bot be controlled by an XBox controller.  That works well.

Now I'd like the bot to be able to play sounds.  To do this, I think I need one coroutine making the `hub.speaker.play_notes` calls while another one runs the main balance loop.  So I added an async main() method around the main loop.  I added an `await` to the last line, ie `await wait(DT)`.  However, that makes the balance bot fall over.  I'm guessing that the context switching for the await takes too long.

So I tried making my own wait implementation instead:

```python
w = StopWatch()
while (w.time() < DT):
    pass
```

The robot does a better job balancing in that case, but it never yields execution so the sounds don't play.  So I tried yielding execution every 100 ms (where await_watch is a StopWatch created outside of the loop):

```python
w = StopWatch()
while (w.time() < DT):
    pass

if (await_watch.time() > 100):
    await wait(1)
    await_watch.reset()
```

This seems to work decently well, though I'm not sure if it balances quite as well as the non-async version.

My solution seems pretty hacky though.  Is there a better way to handle this?

Thanks!

**Comment by dlech** [comment]
Can you share the full Python script?

---

> ```python
> w = StopWatch()
> while (w.time() < DT):
>     pass
> ```

This makes a blocking timer because there is no `await` in the loop, so this can cause problems in a multitasking program.

---

It doesn't look like it is documented, but `run_task()` has a builtin time delay in the loop. The default is 10 ms. So this could be throwing things off since the original script was tuned for a 5ms loop.

For this particular application, it might be best to disable the main loop delay by setting it to 0. Maybe something like this?

```python
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor, ColorSensor
from pybricks.parameters import Port, Direction
from pybricks.tools import wait, StopWatch
from pybricks.geometry import Axis

# Initialize motors.
left = Motor(Port.C, Direction.COUNTERCLOCKWISE)
right = Motor(Port.D)
left.reset_angle(0)
right.reset_angle(0)

# Initialize hub and color sensor.
# You can also use the TechnicHub and/or the ColorDistanceSensor
# instead.
hub = InventorHub()
sensor = ColorSensor(Port.A)


async def balance():
    # Initialize position buffer for speed calculation.
    DT = 5
    window = 300 // DT
    buf = [0] * window
    idx = 0
    angle = 0
    DRIVE_SPEED = 300
    PAUSE = 5000

    # Timer to generate reference position.
    watch = StopWatch()

    while True:
        # Get angular rate and estimated angle.
        rate = hub.imu.angular_velocity(Axis.Y)
        angle += rate * DT / 1000

        # Get motor position.
        position = (left.angle() + right.angle()) / 2

        # Calculate motor speed.
        speed = (position - buf[idx]) / (window * DT) * 1000
        buf[idx] = position
        idx = (idx + 1) % window

        # Calculate reference position, which just grows linearly with
        # time.
        reference = -max(watch.time() - PAUSE, 0) / 1000 * DRIVE_SPEED

        # Calculate duty cycle.
        diff = position - reference
        duty = 0.018 * rate + 19 * angle + 0.45 * diff + 0.16 * speed

        # Account for battery level and type.
        duty *= 7200 / hub.battery.voltage()

        # Calculate steering.
        reflection = await sensor.reflection()
        steering = (reflection - 28) * 0.6

        # Apply duty cycle for balancing and steering.
        left.dc(duty + steering)
        right.dc(duty - steering)

        await wait(DT)

async def sound():
    # TODO - replace with actual code
    while True:
        await wait(1)

async def main():
    await multitask(balance(), sound())

run_task(main(), loop_time=0)
```

**Comment by dsplaisted** [comment]
Thanks.  I tried `looptime=0` and `loop_time=1` with the simple wait statement, but the robot would still appear to try to balance but fall over.  Here's my current code, which more or less works: https://gist.github.com/dsplaisted/b4a4032e1115b80e73c5c36a9baae7d8

**Reply by dlech** [reply]
Hmm... looks like a firmware bug.

**Reply by dlech** [reply]
FYI, should be working better now with latest builds starting with build-3491.

https://nightly.link/pybricks/pybricks-micropython/workflows/build/master

**Reply by dsplaisted** [reply]
Thanks a lot!

**Reply by dlech** [reply]
Another FYI, the first fix wasn't quite right, resulting in crashes when `loop_time=0`. That should be fixed in build-3501 and newer.

**Reply by dsplaisted** [reply]
Hi, I finally got around to installing a nightly firmware to test this (build 3544).  When I tried my program, I got `TypeError: extra keyword arguments given` for this line:

```
run_task(main(), loop_time=0)
```

After removing `loop_time=0` (and switching back to `await wait(DT)` instead of my workaround code), the robot seems to balance well.  Is the `loop_time` argument no longer supported or needed?

Thanks!

**Reply by BertLindeman** [reply]
updated in [build 3510](https://github.com/pybricks/pybricks-micropython/actions/runs/10597001287) 

Commit [Drop run_loop arg](https://github.com/pybricks/pybricks-micropython/commit/261663efcb79bfc86de72df2589aa1d9d8c769ff)

**Reply by dsplaisted** [reply]
Thanks for the help folks!  We used this for my son to tell a fun children's story at our church: https://youtu.be/_-ntSqQ62Ps

**Reply by laurensvalk** [reply]
That's awesome, thanks for sharing!
